types/src/transaction/authenticator.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::account_address::{AccountAddress, AuthenticationKey};
+use libra_crypto::{
+    ed25519::{Ed25519PublicKey, Ed25519Signature},
+    multi_ed25519::{MultiEd25519PublicKey, MultiEd25519Signature},
+    traits::{Signature, VerifyingKey},
+    HashValue,
+};
+
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use std::convert::TryFrom;
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum TransactionAuthenticator {
+    /// Single signature
+    Ed25519 {
+        public_key: Ed25519PublicKey,
+        signature: Ed25519Signature,
+    },
+    /// K-of-N multisig
+    MultiEd25519 {
+        public_key: MultiEd25519PublicKey,
+        signature: MultiEd25519Signature,
+    },
+    // ... add more schemes here
+}
+
+impl TransactionAuthenticator {
+    pub fn edd25519(public_key: Ed25519PublicKey, signature: Ed25519Signature) -> Self {
+        Self::Ed25519 {
+            public_key,
+            signature,
+        }
+    }
+
+    pub fn multi_edd25519(
+        public_key: MultiEd25519PublicKey,
+        signature: MultiEd25519Signature,
+    ) -> Self {
+        Self::MultiEd25519 {
+            public_key,
+            signature,
+        }
+    }
+
+    // TODO: can we derive this automatically from the enum?
+    /// Unique identifier for the signature scheme
+    pub fn scheme_id(&self) -> u8 {
+        match self {
+            Self::Ed25519 { .. } => 0,
+            Self::MultiEd25519 { .. } => 1,
+        }
+    }
+
+    pub fn verify_signature(&self, message: &HashValue) -> Result<()> {
+        match self {
+            Self::Ed25519 {
+                public_key,
+                signature,
+            } => public_key.verify_signature(message, signature),
```

Comment:
Both implement the `Signature` trait, so better use 
`signature.verify(message, public_key)` for consistency.

Reply:
Also, as all of them are `VerifyingKeys`, do you need to distinguish between schemes? or just ensure the enum is correct and then always  `signature.verify`?